### ISSUE TITLE
What's the point of yaml-cpp-targets?

### ISSUE THREAD

**Issue by tambry** [jbeder/yaml-cpp#439]
Currently I have yaml-cpp as a git submodule, that gets added using add_subdirectory to my own project. Calling find_package(yaml-cpp) on CMake 3.7 generates the following error:
```
The file

   .../build/3rdparty/yaml-cpp/yaml-cpp-targets.cmake

  was generated by the export() command.  It may not be used as the argument
  to the include() command.  Use ALIAS targets instead to refer to targets by
  alternative names.
```

After going through the yaml-cpp CMakeLists.txt I'm not quite sure why this file is needed to be generated. Generating this file through the export command generates the error shown above when being built inside another project (ie. as a submodule).  
Note that there's also a SO question [here](http://stackoverflow.com/questions/37204734/correctly-set-cmake-variables-when-adding-yaml-cpp-to-existing-project), that has an answer that is not ideal at all for my usage - I want to include it as a submodule and then build it.

Is this really needed? Is there any way I can workaround this problem, besides simply disabling the generation of that file?

**Comment by tambry** [comment]
I might also might note that disabling the include will fix the problem and seems to cause no other issues.

**Comment by jbeder** [comment]
I don't know; I just accepted a patch request. I don't use find_package, so I just depended on others to make sure it works.

If you can figure out why it's there and update it (ideally rather than simply disabling it), that'd be awesome!

**Comment by tambry** [comment]
After reading [the CMake wiki page on exporting/importing](https://cmake.org/Wiki/CMake/Tutorials/Exporting_and_Importing_Targets#Exporting_from_a_Build_Tree), I'm still quite confused as to why it's there. CMake wiki brings one example of it allowing compiling the project for running on the host machine (ie. special generator or something), that can be run then and then the rest of the project can be cross-compiled from something else, from what I understand.

Would you be fine with adding an option to disable that include? (keeping the current behaviour by default, but allowing to turn it off)

**Comment by jbeder** [comment]
@paulnovo I think you authored the original change a397ad29252b60f0549a4edb389a46a37577cd02. What do you think?

**Comment by sergiud** [comment]
The point of `yaml-cpp-targets.cmake` is that it creates an imported target `yaml-cpp` which ideally would allow to easily link to yaml-cpp using `target_link_libraries` (and no additional commands) after a `find_package` call. However, the export mechanism as currently implemented by yaml-cpp is not relocatable (as every path is kept absolute) making the `yaml-cpp-config.cmake` practically unusable.

@tambry The line should **not** be removed. Please look up in the [CMake documentation](https://cmake.org/cmake/help/v3.0/manual/cmake-packages.7.html#creating-packages) on how to correctly create CMake packages. The [CMakePackageConfigHelpers](https://cmake.org/cmake/help/v2.8.12/cmake.html#module:CMakePackageConfigHelpers) module required for this is available since CMake 2.8.8.

**Comment by sergiud** [comment]
I can provide a fix for this config mess unless someone is already working on this.

**Comment by tambry** [comment]
@sergiud I don't think anyone is working on this yet. At least I'm not and don't see anyone else on the network graph that is.

Linking part is good except if I need to find the include directory. Sure, I got my own project, it will know the directory, but if I have another smaller project as a submodule that also uses yaml-cpp, then it's hard to know the exact include directory. Calling `find_package(yaml-cpp)` in such a situation would be the solution, but currently that's impossible due to the slight mess.

**Comment by sergiud** [comment]
@tambry I see that you actually include yaml-cpp using `add_subdirectory` in addition to calling `find_package (yaml-cpp)`. This is incorrect. Either you call `add_subdirectory` and link to the yaml-cpp target directly, or you use the ExternalProject CMake module in combination with `find_package`. The latter being the recommended method.

**Comment by sergiud** [comment]
@tambry Please test #446.

**Comment by paulnovo** [comment]
Yes, @tambry, you need to use ExternalProject_Add. This should currently work fine, I use yaml-cpp this way. If that works for you, I think we can close this issue, and consider #446 separately.

@sergiud I wouldn't consider the current packaging a mess, it was the recommended way to do this when I implemented this code, admittedly a long time ago, and it works with old (getting very old) versions of CMake.

**Comment by sergiud** [comment]
@paulnovo Sorry for the wording. The reason for calling the current package config a mess is that `yaml-cpp-config.cmake` in most distributions that package yaml-cpp is broken. Specifically, the yaml-cpp imported target often points to non-existent shared objects as the packages are build in a directory different from the install location. Take #362 for example. This is still the case on an update-to-date Arch Linux.

**Comment by paulnovo** [comment]
@sergiud I can have a look at what Arch is doing, but Debian (which I packaged) uses this without problems. Just to clarify, currently, yaml-cpp creates two yaml-cpp-config.cmake files. One for the [Build directory](https://github.com/jbeder/yaml-cpp/blob/master/CMakeLists.txt#L325-L326), with correct paths to the build directory. And one that is [installed](https://github.com/jbeder/yaml-cpp/blob/master/CMakeLists.txt#L336-L337), with correct paths to the install directories.

BTW, #362 seems to be fixed by me in #344, with the code we are discussing here.

**Comment by sergiud** [comment]
@paulnovo The issue is still present in Arch. Building targets that use `yaml-cpp-config.cmake` fail with the error message from #362.

**Comment by paulnovo** [comment]
I don't know much about Arch packaging, but I am looking at the [PKGBUILD file](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/yaml-cpp). I looks like line 27 runs make install, which _should_ install the correct yaml-cpp-config.cmake, yaml-cpp-config-version.cmake, and yaml-cpp-targets.cmake. Lines 29-31 appear to overwrite these files with the ones from the Build directory. Is there a way you could test removing lines 29-31?

**Comment by sergiud** [comment]
@paulnovo The PKGBUILD seems to be fine. The problem however is not in `yaml-cpp-config.cmake`, but in `yaml-cpp-targets.cmake`. The `yaml-cpp` target here points to the `/build/yaml-cpp/src/yaml-cpp-release-0.5.3/libyaml-cpp.so.0.5.3` shared object which does not exist on the host system. The path seems to be used by the Arch build farm. This is with yaml-cpp installed using the package manager. I do not compile yaml-cpp myself.

**Comment by tambry** [comment]
Sorry for the late response.

@sergiud Since you mentioned that you use ExternalProject with yaml-cpp, could you provide an example on how to do that?

I currently have the following:
```cmake
ExternalProject_Add(yaml-cpp GIT_REPOSITORY "git@github.com:jbeder/yaml-cpp.git" GIT_TAG "147d909fe62efc2bfd2bc975785976664b5f79e4")
find_package(yaml-cpp REQUIRED)
```
but the `find_package` fails with:
```
By not providing "Findyaml-cpp.cmake" in CMAKE_MODULE_PATH this project has
asked CMake to find a package configuration file provided by "yaml-cpp",
but CMake did not find one.

...
```

**Comment by tambry** [comment]
Following [this article](https://crascit.com/2015/07/25/cmake-gtest/) I managed to get it working... sort of.

So, I created a sub-CMakeLists.txt file that gets called by my main one and that checks if it can find yaml-cpp using `find_package()`. If not found, it copies another small CMakeLists.txt file to `build/3rdparty/yaml-cpp`, that runs `ExternalProject_Add` to download yaml-cpp, then runs CMake generate on that directory and then lets CMake build yaml-cpp. Then I try to find the package again and re-set the `YAML_CPP_LIBRARIES` and `YAML_CPP_INCLUDE_DIR` as accessible by the parent scope, so I can use them in my main CMakeLists.txt file. And finally then I try do my stuff and include and link the things in those variables.  
But that doesn't seem to work, as the link thingy simply solves to `yaml-cpp.lib`, which obviously doesn't exist in the directory where my project itself is located. Also, the approach above makes the configure step take multiple minutes...

Maybe I'm doing something wrong, but I think I'll stick to hard-coding `YAML_CPP_LIBRARIES` and `YAML_CPP_INCLUDE_DIR`, and adding yaml-cpp as a git submodule that I can easily manage, for the meantime.
